src/behavior.rs

```diff
@@ -59,50 +91,108 @@ pub trait Behavior<Pane> {
     #[allow(clippy::fn_params_excessive_bools)]
     fn tab_ui(
         &mut self,
-        tiles: &Tiles<Pane>,
+        tiles: &mut Tiles<Pane>,
         ui: &mut Ui,
         id: Id,
         tile_id: TileId,
-        active: bool,
-        is_being_dragged: bool,
+        state: TabState,
     ) -> Response {
         let text = self.tab_title_for_tile(tiles, tile_id);
+        let close_btn_size = Vec2::splat(ui.spacing().icon_width);
+        let close_btn_left_padding = 4.0;
         let font_id = TextStyle::Button.resolve(ui.style());
         let galley = text.into_galley(ui, Some(false), f32::INFINITY, font_id);
 
         let x_margin = self.tab_title_spacing(ui.visuals());
-        let (_, rect) = ui.allocate_space(vec2(
-            galley.size().x + 2.0 * x_margin,
-            ui.available_height(),
-        ));
-        let response = ui.interact(rect, id, Sense::click_and_drag());
+        let (_, tab_rect) = match state.closable {
+            true => ui.allocate_space(vec2(
+                galley.size().x + close_btn_size.x + close_btn_left_padding + (2.0 * x_margin),
+                ui.available_height(),
+            )),
+            false => ui.allocate_space(vec2(
+                galley.size().x + (2.0 * x_margin),
+                ui.available_height(),
+            )),
+        };
+
+        let tab_response = ui.interact(tab_rect, id, Sense::click_and_drag());
 
         // Show a gap when dragged
-        if ui.is_rect_visible(rect) && !is_being_dragged {
-            let bg_color = self.tab_bg_color(ui.visuals(), tiles, tile_id, active);
-            let stroke = self.tab_outline_stroke(ui.visuals(), tiles, tile_id, active);
-            ui.painter().rect(rect.shrink(0.5), 0.0, bg_color, stroke);
+        if ui.is_rect_visible(tab_rect) && !state.is_being_dragged {
+            let bg_color = self.tab_bg_color(ui.visuals(), tiles, tile_id, state.active);
+            let stroke = self.tab_outline_stroke(ui.visuals(), tiles, tile_id, state.active);
```

Comment:
let's pass the whole state in to these functions

```suggestion
            let bg_color = self.tab_bg_color(ui.visuals(), tiles, tile_id, state);
            let stroke = self.tab_outline_stroke(ui.visuals(), tiles, tile_id, state);
```